From 82afc0c9bab06f8ed9b4daf97703444451aeabd6 Mon Sep 17 00:00:00 2001
From: Krystian Plata <k.plata@metrological.com>
Date: Tue, 27 Jul 2021 12:11:51 +0000
Subject: [PATCH] BCMCZ-354 Position query handling in pcmsink.

This commit adds handling for a position query bytes format.
This information can be used by an audio only client, to determine
the amount of frames played by the pipeline.

Change-Id: I4ccec8f548690d86d5a4873c4f65f8e146a2e508
---
 reference/pcmsink/src/gst_pcm_sink.c | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/reference/pcmsink/src/gst_pcm_sink.c b/reference/pcmsink/src/gst_pcm_sink.c
index b5ec1ff..777945a 100644
--- a/reference/pcmsink/src/gst_pcm_sink.c
+++ b/reference/pcmsink/src/gst_pcm_sink.c
@@ -1045,7 +1045,20 @@ static gboolean gst_brcm_pcm_sink_query (GstElement *element, GstQuery *query) {
         gst_query_set_latency(query, FALSE, 0, 10*1000*1000);
         return TRUE;
 
-      case GST_QUERY_POSITION:
+      case GST_QUERY_POSITION:{
+        GstFormat queryFormat;
+        gst_query_parse_position(query, &queryFormat, NULL);
+        if (sink->audio_playback && queryFormat == GST_FORMAT_BYTES) {
+            NEXUS_SimpleAudioPlaybackStatus playbackStatus;
+            NEXUS_Error rc = NEXUS_SimpleAudioPlayback_GetStatus(sink->audio_playback, &playbackStatus);
+            if (rc != NEXUS_SUCCESS) {
+                GST_ERROR("NEXUS_SimpleAudioPlayback_GetStatus failed (rc=%d)", rc);
+            } else {
+                gst_query_set_position(query, GST_FORMAT_BYTES, playbackStatus.playedBytes);
+                return TRUE;
+            }
+        } // fall-through!
+      }
       case GST_QUERY_DURATION:
       case GST_QUERY_SEEKING:
       case GST_QUERY_RATE:
-- 
2.7.4

